src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+}
+
+bool strx_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				void *p_result, void *p_min_len,
+				void *p_max_len)
+{
+	if (!strx_start_decode(pp_payload, p_payload_end, p_elem_count, p_result,
+				p_min_len, p_max_len)) {
+		return false;
+	}
+	cbor_string_type_t *p_str_result = (cbor_string_type_t *)p_result;
+	(*pp_payload) += p_str_result->len;
+	return true;
+}
+
+
+bool list_start_decode(uint8_t const **const pp_payload,
+			uint8_t const *const p_payload_end,
+			size_t *const p_elem_count,
+			size_t *p_result, size_t min_num, size_t max_num)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (major_type != CBOR_MAJOR_TYPE_LIST
+		&& major_type != CBOR_MAJOR_TYPE_MAP) {
+		cbor_decode_trace();
+		return false;
+	}
+	if (!uint32_decode(pp_payload, p_payload_end, p_elem_count,
+			p_result, &min_num, &max_num)) {
+		cbor_decode_trace();
+		return false;
+	}
+	return true;
+}
+
+
+bool primx_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				uint8_t *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t major_type = **pp_payload >> 5;
+	uint32_t val;
+
+
+	if (major_type != CBOR_MAJOR_TYPE_PRIM) {
+		/* Value to be read doesn't have the right type. */
+		cbor_decode_trace();
+		return false;
+	}
+	if (!uint32_decode(pp_payload, p_payload_end, p_elem_count,
+			&val, p_min_result, p_max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	if (p_result != NULL) {
+		*p_result = val;
+	}
+	return true;
+}
+
+
+bool boolx_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				bool *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t min_result = *(uint8_t *)p_min_result + 20;
+	uint8_t max_result = *(uint8_t *)p_max_result + 20;
+
+	if (!primx_decode(pp_payload, p_payload_end, p_elem_count,
+			(uint8_t *)p_result, &min_result, &max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	(*p_result) -= 20;
+	return true;
+}
+
+
+bool double_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				double *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	uint8_t major_type = **pp_payload >> 5;
+
+	if (major_type != CBOR_MAJOR_TYPE_PRIM) {
+		/* Value to be read doesn't have the right type. */
+		cbor_decode_trace();
+		return false;
+	}
+	if (!value_extract(pp_payload, p_payload_end, p_elem_count, p_result,
+			sizeof(*p_result))) {
+		cbor_decode_trace();
+		return false;
+	}
+
+	if (!COMP_PTRS(double, p_result, p_min_result, p_max_result)) {
+		cbor_decode_trace();
+		return false;
+	}
+	return true;
+}
+
+
+bool any_decode(uint8_t const **const pp_payload,
+				uint8_t const *const p_payload_end,
+				size_t *const p_elem_count,
+				void *p_result, void *p_min_result,
+				void *p_max_result)
+{
+	if (p_result != NULL) {
```

Comment:
Make it clearer that this is an assert (should not happen)

Reply:
Added a cbor_decode_assert() macro for this and a few other similar cases.